File: shared/isDuplicateCasting.py
import json
import os
import fcntl
from difflib import SequenceMatcher
from shared.normalize import normalize

DB_PATH = os.path.join(os.path.dirname(__file__), '..', 'shared', 'seen_castings_texts.json')
SIMILARITY_THRESHOLD = 0.90

def is_duplicate_casting(text='', ocr_text=''):
    normalized_text = normalize(text or '')
    normalized_ocr = normalize(ocr_text or '')
    combined_normalized = normalized_text + ' | ' + normalized_ocr

    # Убедимся, что файл существует
    if not os.path.exists(DB_PATH):
        with open(DB_PATH, 'w', encoding='utf-8') as f:
            json.dump([], f)

    with open(DB_PATH, 'r+', encoding='utf-8') as db_file:
        try:
            fcntl.flock(db_file, fcntl.LOCK_EX)

            try:
                seen = json.load(db_file)
            except json.JSONDecodeError:
                print("⚠️ Повреждён JSON. Начинаем с пустого списка.")
                seen = []

            # 1️⃣ Прямая проверка
            if combined_normalized in seen:
                print("🔁 Найден дубликат (точное совпадение)")
                return True

            # 2️⃣ Fuzzy проверка
            for entry in seen:
                ratio = SequenceMatcher(None, entry, combined_normalized).ratio()
                if ratio >= SIMILARITY_THRESHOLD:
                    print(f"⚠️ Fuzzy дубликат найден: совпадение {round(ratio * 100)}%")
                    return True

            # 💾 Добавление
            seen.append(combined_normalized)
            db_file.seek(0)
            db_file.truncate()
            json.dump(seen[-20:], db_file, ensure_ascii=False, indent=2)

        finally:
            fcntl.flock(db_file, fcntl.LOCK_UN)

    return False

File: shared/isDuplicateCasting.js
const fs = require('fs');
const path = require('path');
const lockfile = require('proper-lockfile');
const { normalize } = require('./normalize');
const fuzz = require('fuzzball');

const DB_PATH = path.join(__dirname, 'seen_castings_texts.json');
const SIMILARITY_THRESHOLD = 90;

async function isDuplicateCasting(text = '', ocrText = '') {
    const normText = normalize(text || '');
    const normOcr = normalize(ocrText || '');
    const entriesToCheck = [normText, normOcr].filter(Boolean);

    if (!fs.existsSync(DB_PATH)) {
        fs.writeFileSync(DB_PATH, JSON.stringify([]));
    }

    let release;
    try {
        // ⛓️ Ставим безопасный лок
        release = await lockfile.lock(DB_PATH, { retries: 3 });

        let seen;
        try {
            seen = JSON.parse(fs.readFileSync(DB_PATH, 'utf-8'));
        } catch (err) {
            console.error('❌ Повреждён JSON файл дубликатов:', err.message);
            seen = [];
        }

        // 1️⃣ Прямое совпадение
        for (const entry of entriesToCheck) {
            if (seen.includes(entry)) {
                console.log("🔁 Найден точный дубликат");
                return true;
            }
        }

        // 2️⃣ Fuzzy сравнение
        for (const entry of seen) {
            for (const check of entriesToCheck) {
                const ratio = fuzz.token_set_ratio(entry, check);
                if (ratio >= SIMILARITY_THRESHOLD) {
                    console.log(`⚠️ Fuzzy дубликат найден: совпадение ${ratio}%`);
                    return true;
                }
            }
        }

        // 💾 Добавляем и сохраняем
        for (const item of entriesToCheck) {
            if (!seen.includes(item)) {
                seen.push(item);
            }
        }

        fs.writeFileSync(DB_PATH, JSON.stringify(seen.slice(-20), null, 2), 'utf-8');
        return false;

    } catch (err) {
        console.error('❌ Ошибка блокировки или обработки файла:', err.message);
        return false;

     } finally {
        if (release) {
            try {
                await release(); // правильный способ закрыть лок и удалить .lock
            } catch (err) {
                console.error("⚠️ Ошибка при снятии лока:", err.message);
            }
        }
    }
}
module.exports = { isDuplicateCasting };